### PR TITLE
gh-105071: add PyUnstable_Exc_PrepReraiseStar to expose except* implementation in the unstable API

### LINKED ISSUE #105071: Add PyUnstable_Exc_PrepReraiseStar
As requested [here](https://github.com/capi-workgroup/problems/issues/40), it would help cython to have the ``_PyExc_PrepReraiseStar`` function exposed in the C API. Since this is an implementation detail (of the ``except*`` construct), it can only be exposed in the unstable API. 

As per the [dev guide](https://devguide.python.org/developer-workflow/c-api/#unstable-c-api), we can still do this in 3.12.

### Linked PRs
* gh-105072
* gh-105095
* gh-105097
* gh-105105

### PR DESCRIPTION
**Pull request by iritkatriel**
I did not rename the internal function, but rather added an API function that calls it. This is because the internal function can make more assumptions about the input than an API function (so the API function raises exceptions on things that the internal function just asserts). 

Fixes #105071.

CC @da-woods .

* Issue: gh-105071

----
:books: Documentation preview :books:: https://cpython-previews--105072.org.readthedocs.build/

### CODE REVIEW

File: Doc/c-api/exceptions.rst [hunk 1]
```diff
@@ -772,6 +772,16 @@ Exception Objects
 
    Set :attr:`~BaseException.args` of exception *ex* to *args*.
 
+.. c:function:: PyObject* PyUnstable_Exc_PrepReraiseStar(PyObject *orig, PyObject *excs)
+
+   Implements part of the interpreter's implementation of :keyword:`!except*`.
```
Comment (erlend-aasland) [inline Doc/c-api/exceptions.rst]:
```suggestion
   Implement part of the interpreter's implementation of :keyword:`!except*`.
```

File: Doc/c-api/exceptions.rst [hunk 2]
```diff
@@ -772,6 +772,16 @@ Exception Objects
...
+   *orig* is the original exception that was caught, and *excs* is the list of
+   the exceptions that need to be raised. This list contains the the unhandled
+   part of orig, if any, as well as the exceptions that were raised from the
```
Comment (erlend-aasland) [inline Doc/c-api/exceptions.rst]:
```suggestion
   part of *orig*, if any, as well as the exceptions that were raised from the
```

File: Doc/c-api/exceptions.rst [hunk 3]
```diff
@@ -772,6 +772,16 @@ Exception Objects
...
+   ``except*`` clauses (so they have a different traceback from *orig*) and
```
Comment (erlend-aasland) [inline Doc/c-api/exceptions.rst]:
```suggestion
   :keyword:`!except*` clauses (so they have a different traceback from *orig*) and
```

**Review by erlend-aasland — COMMENTED** [review]
Doc formatting nits. Otherwise: lgtm, as far as I can see.

File: Objects/exceptions.c [hunk 4]
```diff
@@ -1438,6 +1441,43 @@ _PyExc_PrepReraiseStar(PyObject *orig, PyObject *excs)
...
+        PyObject *exc = PyList_GET_ITEM(excs, i);
+        if (exc == NULL || !(PyExceptionInstance_Check(exc) || Py_IsNone(exc))) {
+            PyErr_Format(PyExc_TypeError,
+                         "item %d of excs is not an exception", i);
+            return NULL;
+        }
+    }
+
+    /* Make sure that orig has something as traceback, in the interpreter
+     * it always does becuase it's a raised exception).
```
Comment (iritkatriel) [inline Objects/exceptions.c]:
```suggestion
     * it always does becuase it's a raised exception.
```

File: Objects/exceptions.c [hunk 5]
```diff
@@ -1438,6 +1441,43 @@ _PyExc_PrepReraiseStar(PyObject *orig, PyObject *excs)
...
+     */
+    PyObject *tb = PyException_GetTraceback(orig);
+    assert(tb == NULL || !Py_IsNone(tb));
```
Comment (iritkatriel) [inline Objects/exceptions.c]:
```suggestion
```

**Review by encukou — APPROVED** [review]
Looks good, thanks!

**Pull request merged**